test/metabase/lib/test_util/generators/filters.cljc

```diff
@@ -0,0 +1,278 @@
+       (< (rand) 0.2)))
+
+(defn- gen-filter:date-binary [column operator]
+  (if (specify-time? column)
+    (operator (lib/with-temporal-bucket column :minute) (gen-datetime:minute))
+    (operator column (gen-date))))
+
+(defn- gen-filter:date-between [column]
+  ;; TODO: Swap the arguments to put the earlier one on the left.
+  (if (specify-time? column)
+    (lib/between (lib/with-temporal-bucket column :minute)
+                 (gen-datetime:minute) (gen-datetime:minute))
+    (lib/between column (gen-date) (gen-date))))
+
+(defn- gen-filter:date [column]
+  ;; - 30% relative date ranges
+  ;; - 20% exclude
+  ;; - 50% before/after/on/between
+  (let [r (rand)]
+    (cond
+      (< r 0.30) (gen-filter:relative-date column)
+      (< r 0.50) (gen-filter:exclude-date column)
+      (< r 0.60) (gen-filter:date-binary column lib/<)
+      (< r 0.70) (gen-filter:date-binary column lib/>)
+      (< r 0.85) (gen-filter:date-binary column lib/=)
+      :else      (gen-filter:date-between column))))
+
+(defn- gen-filter:datetime [column]
+  ;; TODO: There's actually no difference right now. Clean this up?
+  (gen-filter:date column))
+
+(defn- gen-filter:generic [column]
+  (when-let [operator (some->> (:operators column)
+                               (remove skipped-operator?)
+                               rand-nth)]
+    (gen-filter-clause column operator)))
+
+(defn- gen-filter:inside [col1 col2]
+  (let [[lat lon] (if (lib.types.isa/latitude? col1)
+                    [col1 col2]
+                    [col2 col1])
+        [lat-min lat-max] (sort (repeatedly 2 gen-latitude))
+        [lon-min lon-max] (sort (repeatedly 2 gen-longitude))]
+    ;; Yes, this is really the argument order for an `:inside` clause.
+    (lib/inside lat lon lat-max lon-min lat-min lon-max)))
+
+(defn- gen-filter:coordinate [column]
+  (let [counterpart? (if (lib.types.isa/latitude? column)
+                       lib.types.isa/longitude?
+                       lib.types.isa/latitude?)
+        counterparts (filter counterpart? *filterable-columns*)]
+    (if (and (seq counterparts)
+             (< (rand) 0.5))
+      ;; If we found a coordinate pair, generate an :inside filter 50% of the time.
+      (gen-filter:inside column (gen.u/choose counterparts))
+      ;; Otherwise, generic filter on the original column.
+      (gen-filter:generic column))))
+
+(defn- ^:private gen-filter* []
+  (let [column (rand-nth *filterable-columns*)]
+    (cond
+      (lib.types.isa/coordinate? column)             (gen-filter:coordinate column)
+      (lib.types.isa/date-without-time? column)      (gen-filter:date column)
+      (isa? (:effective-type column) :type/DateTime) (gen-filter:datetime column)
+      :else
+      (let [result (gen-filter:generic column)]
+        ;; Sometimes we pick a column with no filter operators, and result is nil. Recur in that case to roll again.
+        (if (= result ::no-operators)
+          (recur)
```

Comment:
Is it possible to get here with a single column in `*filterable-columns*`? Or, I guess, any number of columns, but by bad luck all return `::no-operators`?

Reply:
That's conceivable, though unlikely. I should at least guard against an infinite loop in testing.

Reply:
Done.